### PR TITLE
[DOC] Packed data

### PR DESCRIPTION
**Pull request by BurdetteLamar**
New page for packed data, full-fledged discussion with many examples.  Will provide link targets for Array#pack, String#unpack, and String#unpack1.  Many examples adapted from the tests.

### CODE REVIEW

File: doc/packed_data.rdoc [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+  if count is zero, no newlines are added;
+  (see {RFC 4648}[https://www.ietf.org/rfc/rfc4648.txt]):
+
+    [""].pack('m')             # => ""
+    ["\0"].pack('m')           # => "AA==\n"
+    ["\0\0"].pack('m')         # => "AAA=\n"
+    ["\0\0\0"].pack('m')       # => "AAAA\n"
+    ["\377"].pack('m')         # => "/w==\n"
+    ["\377\377"].pack('m')     # => "//8=\n"
+    ["\377\377\377"].pack('m') # => "////\n"
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc:503]: Taken straight from the tests;  which should be eliminated?

File: doc/packed_data.rdoc [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+
+- \Method Array#pack:
+  Formats each element in array +self+ into a binary string;
+  returns that string.
+- \Method String#unpack:
+  Extracts data from string +self+,
+  forming objects that become the elements of a new array;
+  returns that array.
+- \Method String#unpack1:
+  Does the same, but the returned array
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.

File: doc/packed_data.rdoc [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+  [65, -1].pack('cC')  # => "A\xFF"
+  "A\xFF".unpack('cC') # => [65, 255]
+
+The string +template+ may contain comments,
+each of which begins with character <tt>'#'</tt>
+and continues up to and including the next following newline:
+
+
+  [0,1].pack("  C  #foo \n  C  ")    # => "\x00\x01"
+  "\0\1".unpack("  C  #foo \n  C  ") # => [0, 1]
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc:44]: Done.

File: doc/packed_data.rdoc [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+
+- Integer +count+ - The directive is to be applied +count+ times:
+
+    [65, 66].pack('C2') # => "AB"
+    [65, 66].pack('C3') # Raises ArgumentError.
+    'AB'.unpack('C2')   # => [65, 66]
+    'AB'.unpack('C3')   # => [65, 66, nil]
+
+  Note: Directives in <tt>%w[A a Z m]</tt> use +count+ differently;
+  see {String Directives}[rdoc-ref:packed_data.md@String+Directives].
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.

File: doc/packed_data.rdoc [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+  [65, 66].pack('C*', buffer: 'foo') # => "fooAB"
+
+Methods String#unpack and String#unpack1 each accept
+an optional keyword argument +offset+ that specifies an offset
+into the string:
+
+  'ABC'.unpack('C*', offset: 1)  # => [66, 67]
+  'ABC'.unpack1('C*', offset: 1) # => 66
+
+Any of the three methods may have a block:
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Done.  (I think this is what you suggested.

File: doc/packed_data.rdoc [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+  (like C <tt>uint64_t</tt>):
+
+    s = [578437695752307201, 17940646550795321087].pack('Q*')
+    # => "\x01\x02\x03\x04\x05\x06\a\b\xFF\xFE\xFD\xFC\xFB\xFA\xF9\xF8"
+    s.unpack('Q*')
+    # => [578437695752307201, 17940646550795321087]
+
+==== Platform-Dependent \Integer Directives
+
+- <tt>'i'</tt> - Platform-dependent width signed integer,
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc:197]: Added.

File: doc/packed_data.rdoc [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+  (see {BER enocding}[https://en.wikipedia.org/wiki/X.690#BER_encoding]):
+
+    s = [1073741823].pack('w*')
+    # => "\x83\xFF\xFF\xFF\x7F"
+    s.unpack('w*')
+    # => [1073741823]
+
+==== Modifiers for \Integer Directives
+
+For directives in <tt>%w[ i I s S l L q Q j J]</tt>,
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.

File: doc/packed_data.rdoc [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+  "\x00\x00\xC0\x7F".unpack('f') # => [NaN]
+
+=== \String Directives
+
+Each string directive specifies the packing or unpacking
+for one byte in the input or output string.
+
+==== Binary \String Directives
+
+- <tt>'A'</tt> - Arbitrary binary string (space padded; count is width):
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Done.

File: doc/packed_data.rdoc [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+    ['foo'].pack('A*')   # => "foo"
+    ['foo'].pack('A2')   # => "fo"
+    ['foo'].pack('A4')   # => "foo "
+    [nil].pack('A')      # => " "
+    [nil].pack('A*')     # => ""
+    [nil].pack('A2')     # => "  "
+    [nil].pack('A4')     # => "    "
+
+    "foo\0".unpack('A')  # => ["f"]
+    "foo\0".unpack('A4') # => ["foo"]
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Done.

File: doc/packed_data.rdoc [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+    "foo ".unpack('a4')  # => ["foo "]
+    "foo".unpack('a4')   # => ["foo"]
+
+- <tt>'Z'</tt> - Same as <tt>'a'</tt>,
+  except that null is added or ignored with <tt>'*'</tt>:
+
+    ["foo"].pack('Z*')   # => "foo\x00"
+    [nil].pack('Z*')     # => "\x00"
+
+    "foo\0".unpack('Z*') # => ["foo"]
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Done.

File: doc/packed_data.rdoc [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+    "\x01\xfe".unpack('h3') # => ["10e"]
+    "\x01\xfe".unpack('h4') # => ["10ef"]
+    "\x01\xfe".unpack('h5') # => ["10ef"]
+
+==== Pointer \String Directives
+
+- <tt>'P'</tt> - Pointer to a structure (fixed-length string):
+
+    s = ['abc'].pack('P')     # => "\xE0O\x7F\xE5\xA1\x01\x00\x00"
+    s.freeze.unpack('P*')     # => ["abc"]
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.

File: doc/packed_data.rdoc [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+    ".".unpack("P")           # => []
+    ("\0" * 1024).unpack("P") # => [nil]
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.  (1024 was from tests.)

File: doc/packed_data.rdoc [hunk 13]
```diff
@@ -0,0 +1,586 @@
...
+  "\x00\x00\xC0\x7F".unpack('f') # => [NaN]
+
+=== \String Directives
+
+Each string directive specifies the packing or unpacking
+for one byte in the input or output string.
+
+==== Binary \String Directives
+
+- <tt>'A'</tt> - Arbitrary binary string (space padded; count is width)
```
Comment (peterzhu2118) [inline doc/packed_data.rdoc:323]: Should there be a period at the end of this line? The next line feels like a new sentence to me.

File: doc/packed_data.rdoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (peterzhu2118) [reply doc/packed_data.rdoc:524]: I would feel it's a bit redundant if it doesn't highlight anything important. Most of the tests is more verbose than the examples we need to give.

File: doc/packed_data.rdoc (hunk 13 above)
Comment (BurdetteLamar) [reply doc/packed_data.rdoc]: Fixed.

**Comment by peterzhu2118** [comment]
I do think the examples are bit too verbose since, for me at least, longer examples require more mental capacity as I need to first understand the code, figure out how the code differs from other examples, and then understand what the example is trying to demonstrate.

On the other hand, it's difficult to suggest which examples to remove as many of them are demonstrating something different.

But I guess it's a tradeoff. Longer examples provide more coverage of corner cases at the cost of the possibility that the reader loses focus halfway. On the other hand, shorter examples have more focus on what you're trying to demonstrate at the cost of completeness.

At the end of the day, I'm not sure what the best option is. I prefer having shorter examples, but I don't have strong opinions on this.

**Comment by BurdetteLamar** [comment]
I'm not wed to these examples, but I don't feel comfortable to winnow them by deciding which should stay/go.  So, from you I request either approval or winnowing.  Or perhaps someone else should look at them?

**Review by peterzhu2118 — APPROVED** [review]
While I think the examples could use some trimming, I don't feel strongly about it (it's a tradeoff of completeness vs. concise). The other parts look good, so I don't block this PR solely based off that.

**Pull request merged**